base/event.jl

```diff
@@ -1,5 +1,67 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+## thread/task locking abstraction
+
+"""
+    AbstractLock
+
+Abstract supertype describing types that
+implement the thread-safe synchronization primitives:
+[`lock`](@ref), [`trylock`](@ref), [`unlock`](@ref), and [`islocked`](@ref).
+"""
+abstract type AbstractLock end
+function lock end
+function unlock end
+function trylock end
+function islocked end
+unlockall(l::AbstractLock) = unlock(l) # internal function for implementing `wait`
+relockall(l::AbstractLock, token::Nothing) = lock(l) # internal function for implementing `wait`
+assert_havelock(l::AbstractLock) = assert_havelock(l, Threads.threadid())
+assert_havelock(l::AbstractLock, tid::Integer) =
+    (islocked(l) && tid == Threads.threadid()) ? nothing : error("concurrency violation detected")
+assert_havelock(l::AbstractLock, tid::Task) =
+    (islocked(l) && tid === current_task()) ? nothing : error("concurrency violation detected")
+assert_havelock(l::AbstractLock, tid::Nothing) = error("concurrency violation detected")
+
+"""
+    NotALock
+
+A struct that pretends to be always locked on the original thread it was allocated on,
+and simply ignores all other interactions.
+This can be used in the place of a real lock to, instead, simply and cheaply assert
+that the operation is only occurring on a single thread.
+And is thus functionally equivalent to allocating a real, recursive lock,
+immediately calling `lock` on it, and then never calling a matching `unlock`,
+except that calling `lock` from another thread will throw a concurrency violation exception.
+"""
+struct NotALock <: AbstractLock
+    ownertid::Int16
+    NotALock() = new(Threads.threadid())
+end
+assert_havelock(l::NotALock) = assert_havelock(l, l.ownertid)
+lock(l::NotALock) = assert_havelock(l)
+unlock(l::NotALock) = assert_havelock(l)
+trylock(l::NotALock) = l.ownertid == Threads.threadid()
+islocked(::NotALock) = true
+
+"""
+    CooperativeLock
+
+An optimistic lock, which can be used cheaply to check for missing
```

Comment:
Likewise, `OptimisticLock`? You don't use the work "cooperative" anywhere in the description of what this does.

Reply:
Optimistic locking is already a standard term for something different. This is a cooperative lock for cooperative threading (what I describe as "trivial" in the docstring).